File: 1605-minimum-number-of-days-to-make-m-bouquets/1605-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int minDays(vector<int>& bloomDay, int m, int k) {
        int n = bloomDay.size();
        int left = 0, right = 1e9, ans = -1;

        while(left <= right) {
            int mid = (right - left)/2 + left;
            int con = 0, buk = 0;

            for(int i = 0; i<n; i++) {
                if(bloomDay[i] <= mid) {
                   con++;
                   if(con >= k) {
                       con = 0;
                       buk++;
                   }
                }
                else {
                    con = 0;
                }
            }

            if(buk >= m) {
                ans = mid;
                right = mid - 1;
            }
            else {
                left = mid + 1;
            }
        }

        return ans;
    }
};

File: 1605-minimum-number-of-days-to-make-m-bouquets/1605-minimum-number-of-days-to-make-m-bouquets.py
class Solution(object):
    def minDays(self, bloomDay, m, k):
        left, right, ans = 0, 10**9, -1

        while left <= right:
            mid = (right + left)/2
            con, buk = 0, 0
            for n in bloomDay:
                if n <= mid:
                    con += 1
                    if con >= k:
                        buk += 1
                        con = 0
                else:
                    con = 0
            
            if buk >= m:
                ans = mid
                right = mid - 1
            else:
                left = mid + 1
        
        return ans
        